pkg/kubelet/cm/cpumanager/cpu_manager.go

```diff
@@ -239,15 +241,68 @@ type reconciledContainer struct {
 	containerID   string
 }
 
-func (m *manager) reconcileState() (success []reconciledContainer, failure []reconciledContainer) {
+func (m *manager) removeStaleState() {
+	// Only once all sources are ready do we attempt to remove any stale state.
+	// This ensures that the call to `m.activePods()` below will succeed with
+	// the actual active pods list.
 	if !m.sourcesReady.AllReady() {
 		return
 	}
+
+	// We grab the lock to ensure that no new containers will grab CPUs while
+	// executing the code below. Without this lock, its possible that we end up
+	// removing state that is newly added by an asynchronous call to
+	// AddContainer() during the execution of this code.
+	m.Lock()
+	defer m.Unlock()
+
+	// We remove stale state very conservatively, only removing *any* state
+	// once we know for sure that we wont be accidentally removing state that
+	// is still valid. Since this function is called periodically, we will just
+	// try again next time this function is called.
+	activePods := m.activePods()
+	if len(activePods) == 0 {
+		// If there are no active pods, skip the removal of stale state.
+		return
+	}
+
+	// Build a list of containerIDs for all containers in all active Pods.
+	activeContainers := make(map[string]struct{})
+	for _, pod := range activePods {
+		pstatus, ok := m.podStatusProvider.GetPodStatus(pod.UID)
+		if !ok {
+			// If even one pod does not have it's status set, skip state removal.
+			return
+		}
+		for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+			containerID, err := findContainerIDByName(&pstatus, container.Name)
+			if err != nil {
+				// If even one container does not have it's containerID set, skip state removal.
```

Comment:
I wonder why state removal cannot be done for the other containers.

Reply:
I'm not sure I understand the concern. This function removes state from *all* containers that don't make it into the `activeContainers` list. If there's a possibility that we might not add a container to this list when we should, then the only thing we can do is exit early and try again another time.

Reply:
I think error from getting containerID for one of InitContainers shouldn't skip all of Containers

Reply:
I'm happy to consider a modification if you can suggest a way to do that. I don't immediately see a way to do it though.

Reply:
How about the following ?
```
		for _, container := range pod.Spec.InitContainers {
			containerID, err := findContainerIDByName(&pstatus, container.Name)
			if err != nil {
				break
			}
			activeContainers[containerID] = struct{}{}
		}
		for _, container := range pod.Spec.Containers {
			containerID, err := findContainerIDByName(&pstatus, container.Name)
			if err != nil {
				// If even one container does not have it's containerID set, skip state removal.
				return
			}
			activeContainers[containerID] = struct{}{}
		}
```
If you think the above is better, I can send out a PR after this PR merges.

Reply:
That won't work though.

If any call to `findContainerIDByName()` fails then the `containerID` of that container won't be added to the `activeContainers` list (even though there might actually be valid state for that container).

At the end of the function we then remove any state *not* part of the `activeContainers` list, which is unacceptable if it was just a sporadic syncing of the pod state.

In any case, this entire problem goes away in the next PR I have open (there will be no need to call out to `findContainerIDByName()` going forward):
https://github.com/kubernetes/kubernetes/pull/84462

Reply:
Glad you have plan for improving this part of code.